src/graph/graph.c

```diff
@@ -943,42 +931,61 @@ void _BulkDeleteEdges(Graph *g, Edge *edges, size_t edge_count) {
-		deletion = deletions[i];
-		assert(GxB_Matrix_Delete(deletion.M, deletion.row, deletion.col) == GrB_SUCCESS);
-	}
-
-	int relationCount = Graph_RelationTypeCount(g);
-	uint deletion_count = array_len(deletions);
-	for(uint i = 0; i < deletion_count; i++) {
-		deletion = deletions[i];
-		GrB_Index src = deletion.row;
-		GrB_Index dest = deletion.col;
+	if(need_update) {
+		GrB_Matrix remaining_mask;
+		GrB_Matrix_new(&remaining_mask, GrB_BOOL, Graph_RequiredMatrixDim(g), Graph_RequiredMatrixDim(g));
+		GrB_Descriptor desc;    // GraphBLAS descriptor.
+		GrB_Descriptor_new(&desc);
+		// Descriptor sets to clear entry according to mask.
+		GrB_Descriptor_set(desc, GrB_MASK, GrB_SCMP);
 
-		// See if source is connected to destination with additional edges.
-		bool connected = false;
-		for(int i = 0; i < relationCount; i++) {
-			R = Graph_GetRelationMatrix(g, i);
-			info = GrB_Matrix_extractElement_BOOL(&connected, R, src, dest);
-			if(info == GrB_SUCCESS) break;
+		// Clear updated output matrix before assignment.
+		GrB_Descriptor_set(desc, GrB_OUTP, GrB_REPLACE);
+
+		/* Calling GrB_Matrix_apply over a matrix and mask, without setting GrB_MASK = GrB_SCMP, or setting
+		 * the desctiptort to GrB_NULL
+		 * GrB_Matrix_apply(res, mask, NULL, GrB_IDENTITY_UINT64, mat, NULL)
+		 * is matrix addition: res = mat + mask.
+		 * When the descriptor set to GrB_MASK = GrB_SCMP, calling
+		 * GrB_Matrix_apply(res, mask, NULL, GrB_IDENTITY_UINT64, mat, desc)
+		 * is matrix substructing: res = mat - mask. */
+
+		for(int r = 0; r < relationCount; r++) {
+			GrB_Matrix mask = masks[r];
+			GrB_Matrix R = Graph_GetRelationMatrix(g, r); // Relation Matrix.
+			GrB_Index nvals;
+			if(mask) {
+				GrB_Matrix M = Graph_GetRelationMap(g, r);  // Relation mapping matrix.
+				GrB_Matrix_nvals(&nvals, mask);
+				// Remove every entry of R and M marked by Mask.
+				// R = R - mask.
+				GrB_Matrix_apply(R, mask, NULL, GrB_IDENTITY_UINT64, R, desc);
+				// M = M - mask.
+				GrB_Matrix_apply(M, mask, NULL, GrB_IDENTITY_UINT64, M, desc);
+				GrB_free(&mask);
+			}
+			// Collect remaining edges. remaining_mask = remaining_mask + R.
+			GrB_Matrix_apply(remaining_mask, R, NULL, GrB_IDENTITY_UINT64, R, NULL);
 		}
 
-		/* There are no additional edges connecting source to destination
-		 * Remove edge from THE adjacency matrix.
-		 * It is OK to remove entries from the adjacency matrix, as we're
-		 * not trying to extract entries from it. */
-		if(!connected) {
-			M = Graph_GetAdjacencyMatrix(g);
-			assert(GxB_Matrix_Delete(M, src, dest) == GrB_SUCCESS);
+		GrB_Matrix adj_matrix = Graph_GetAdjacencyMatrix(g);
+		GrB_Matrix t_adj_matrix = _Graph_Get_Transposed_AdjacencyMatrix(g);
+		// To calculate edges to delete, remove all the remaining edges from "The" adjency matrix.
+		// remaining_mask = adj_matrix - remaining_mask.
+		GrB_Matrix_apply(remaining_mask, remaining_mask, NULL, GrB_IDENTITY_UINT64, adj_matrix, desc);
```

Comment:
Isn't this step redundant 

Can't we do with just:
GrB_Matrix_apply(adj_matrix, remaining_mask, NULL, GrB_IDENTITY_UINT64, adj_matrix, GrB_NULL);

BTW for GraphBLAS replace all NULL with GrB_NULL.

Reply:
correct. done.